src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,245 @@
+    # what does that mean? either:
+    #
+    # 1. contrasts.levels == levels (best case)
+    # 2. data levels missing from contrast: would generate empty/undefined rows. 
+    #    better to filter data frame first
+    # 3. contrast levels missing from data: would have empty columns, generate a
+    #    rank-deficient model matrix.
+    c_levels = oftype(levels, get(contrasts.levels, levels))
+    mismatched_levels = symdiff(c_levels, levels)
+    isempty(mismatched_levels) || error("contrasts levels not found in data or vice-versa: $mismatched_levels.\nData levels: $levels.\nContrast levels: $c_levels")
+
+    n = length(c_levels)
+    n == 0 && error("empty set of levels found (need at least two to compute contrasts).")
+    n == 1 && error("only one level found: $(c_levels[1]). need at least two to compute contrasts.")
+    
+    # find index of base level. use contrasts.base, then default (1).
+    baseind = isnull(contrasts.base) ?
+              1 :
+              findfirst(c_levels, convert(eltype(levels), get(contrasts.base)))
+    baseind > 0 || error("base level $(contrasts.base) not found in levels $c_levels.")
+
+    tnames = termnames(contrasts, c_levels, baseind)
+
+    mat = contrasts_matrix(contrasts, baseind, n)
+
+    ContrastsMatrix(mat, tnames, c_levels, contrasts)
+end
+
+# Methods for constructing ContrastsMatrix from data. These are called in
+# ModelFrame constructor and setcontrasts!.
+
+ContrastsMatrix(C::AbstractContrasts, v::PooledDataArray) = ContrastsMatrix(C, levels(v))
+# instantiate Type{C<:AbstractContrasts}
+ContrastsMatrix{C <: AbstractContrasts}(c::Type{C}, col::PooledDataArray) = ContrastsMatrix(c(), col)
+# given an existing ContrastsMatrix, check that all of the levels present in the
+# data are present in the contrasts. Note that this behavior is different from the
+# ContrastsMatrix constructor, which requires that the levels be exactly the same.
+# This method exists to support things like `predict` that can operate on new data
+# which may contain only a subset of the original data's levels. Checking here
+# (instead of in `modelmat_cols`) allows an informative error message.
+ContrastsMatrix(c::ContrastsMatrix, col::PooledDataArray) =
+    isempty(setdiff(levels(col), c.levels)) ?
+    c :
+    error("there are levels in data that are not in ContrastsMatrix: $(setdiff(levels(col), c.levels))\n  Data levels: $(levels(col))\n  Contrast levels $(c.levels)")
+
+function termnames(C::AbstractContrasts, levels::Vector, baseind::Integer)
+    not_base = [1:(baseind-1); (baseind+1):length(levels)]
+    levels[not_base]
+end
+
+nullify(x::Nullable) = x
+nullify(x) = Nullable(x)
+
+# Making a contrast type T only requires that there be a method for
+# contrasts_matrix(T, v::PooledDataArray). The rest is boilerplate.
+for contrastType in [:TreatmentContrasts, :SumContrasts, :HelmertContrasts]
+    @eval begin
+        type $contrastType <: AbstractContrasts
+            base::Nullable{Any}
```

Comment:
I think you should restrict the types of the fields. `levels` should be `Nullable{Vector{T}}` with `T` a type parameter, and `base` should be a `Nullable{Int}` (or `Nullable{T}`). Apart from removing type instability, that should help making it clearer which types are expected in the rest of the code.